### PR TITLE
Limit size of loaded file systems

### PR DESCRIPTION
**Pull request by JosJuice**
If a disc image is corrupt in a specific way, Dolphin will try to allocate a lot of memory, making it crash. (https://github.com/dolphin-emu/dolphin/commit/2d5d5fa83e9ccb32252756fc0390740c952121b1#commitcomment-11147057)

To avoid that, this change adds an artificial limit for the size of file systems that Dolphin will try to load.

### CODE REVIEW

File: Source/Core/DiscIO/FileSystemGCWii.cpp [hunk 1]
```diff
@@ -270,6 +271,18 @@ void CFileSystemGCWii::InitFileSystem()
 	if (!Root.IsDirectory())
 		return;
 
+	if (Root.m_FileSize > 2 * 1024 * 1024)
```
Comment (lioncash) [inline Source/Core/DiscIO/FileSystemGCWii.cpp]: The limit size should have its own named constant.

File: Source/Core/DiscIO/FileSystemGCWii.cpp [hunk 2]
```diff
@@ -270,6 +271,19 @@ void CFileSystemGCWii::InitFileSystem()
 	if (!Root.IsDirectory())
 		return;
 
+	static const u32 ARBITRARY_FILE_SYSTEM_SIZE_LIMIT = 2 * 1024 * 1024;
+	if (Root.m_FileSize > ARBITRARY_FILE_SYSTEM_SIZE_LIMIT)
+	{
+		// Without this check, Dolphin may crash by trying to allocate too much memory
+		// when loading the filesystems of certain corrupt disc images. 12 bytes
+		// (the size of a file entry) times 2 * 1024 * 1024 is equal to the size of
+		// main GC/Wii RAM (24 MiB), and no file system should use anywhere near
```
Comment (bb010g) [inline Source/Core/DiscIO/FileSystemGCWii.cpp]: Perhaps move this explanation about the 2 * 1024 * 1024 size to `ARBITRARY_FILE_SYSTEM_SIZE_LIMIT`?

Comment (JosJuice) [inline Source/Core/DiscIO/FileSystemGCWii.cpp]: I don't want to move the whole comment to `ARBITRARY_FILE_SYSTEM_SIZE_LIMIT` because it's mainly about the check, and I don't want to only move part of it because it would be awkward to write and read this explanation in two separated parts.

File: Source/Core/DiscIO/FileSystemGCWii.cpp [hunk 3]
```diff
@@ -270,6 +271,19 @@ void CFileSystemGCWii::InitFileSystem()
 	if (!Root.IsDirectory())
 		return;
 
+	static const u32 ARBITRARY_FILE_SYSTEM_SIZE_LIMIT = 2 * 1024 * 1024;
+	if (Root.m_FileSize > ARBITRARY_FILE_SYSTEM_SIZE_LIMIT)
```
Comment (degasus) [inline Source/Core/DiscIO/FileSystemGCWii.cpp]: To make it less hacky, you could limit by m_rVolume->GetSize() / 0xC

Comment (JosJuice) [inline Source/Core/DiscIO/FileSystemGCWii.cpp]: Unfortunately, GetSize() isn't always correct. Some compressed formats return the space the disc image takes after compression, and I think VolumeDirectory always returns 0.

**Comment by shuffle2** [comment]
Allocating >= 4GB of ram is no problem, that is what virtual memory is for.
Why not just see if the allocation failed and then fail?
ARBITRARY_FILE_SYSTEM_SIZE_LIMIT seems pretty small.

**Comment by JosJuice** [comment]
After allocating that much memory, Dolphin will try to write to it all. If the user doesn't have enough RAM, that's at least going to be very slow.

How do I check if the allocation fails? I thought exception handling was disabled in Dolphin...

I've increased the limit from 24 MiB to 120 MiB.

**Comment by shuffle2** [comment]
I still dont think putting a limit on file size (below disc size) is a good idea. there is undoubtedly some stupid game out there that has just one giant compressed file or something.  
Other approaches:
 * verify disc hash structure (present on wii discs)
 * sanity-check FST values (e.g. go through the entries and see if total size is less than disc size, no entries overlap, etc)

**Comment by shuffle2** [comment]
Actually, looking closer, this is just "filesize" for Root...which is the count of entries, nothing to do with filesize.
The FST size should be given elsewhere on the disc. if root.size * entry_size != fst_size then you know it's broken, no?

**Comment by JosJuice** [comment]
Verifying disc hashes will catch unintentionally malformed disc images but not intentionally malformed ones. Is that good enough, and will such a change be accepted in stable? Will such a change even be made in time for stable?

The allocation happens before I can read the FST entries. I suppose it's possible to skip the `reserve` and just allocate memory gradually, but I'd like to have a solution that doesn't break with PR #2820.

No, because `fst_size == root.size * entry_size + name_table_size`. Besides, I can't rely on `fst_size` being correct.

In short, your suggestions make sense (except for `root.size * entry_size != fst_size` specifically) and would catch most problems. However, just having a limit will catch all allocation problems, and it will not cause any problems with valid games. Games need to load FSTs into RAM to run, and 120 MiB is larger than GC/Wii RAM, so any file systems that this PR won't load won't work on a real console and can't run in Dolphin.

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by aserna3** [comment]
Ping to get this merged into stable. This will clear a blocker.

**Comment by Armada651** [comment]
@aserna3 Sure, but first get this merged into master.

**Comment by JosJuice** [comment]
@Armada651 Do you mean that we should get this PR merged or that I should make a new PR for master? This PR is for stable.

**Comment by Armada651** [comment]
Oh I didn't see this was for stable, so should this be picked for master too?

**Comment by JosJuice** [comment]
Sure, if you're fine with doing cherry-picks to master. You said you weren't going to do any when we were talking about #2815, but if that's changed, then go ahead. (After the PR is merged, that is)

**Comment by Armada651** [comment]
@JosJuice I'd still prefer not to.

**Comment by JosJuice** [comment]
Making PRs for stable gets too messy with our release process. I've remade this PR for master: #3206
